fixed black background of transparent png thumbnail for older gd version

raffaelj opened the pull request:
fixes #236

claviska commented on src/claviska/SimpleImage.php:
```diff
@@ -579,15 +578,47 @@ public function crop($x1, $y1, $x2, $y2) {
-      break;
-    default:
+    }
+    elseif ($this->mimeType == 'image/gif' && GD_BUNDLED === 0){
```
Same line as the previous closing bracket, for consistency :)

claviska commented on src/claviska/SimpleImage.php:
```diff
@@ -579,15 +578,47 @@ public function crop($x1, $y1, $x2, $y2) {
+        
+        $newImage = imagecreatetruecolor($width, $height);
+        
+        $trnprt_indx = imagecolortransparent($this->image);
```
Please use camelCase vars for consistency.

claviska commented on src/claviska/SimpleImage.php:
```diff
@@ -579,15 +578,47 @@ public function crop($x1, $y1, $x2, $y2) {
+        
+        $this->image = $newImage;
+    }
+    else {
```
Same line as bracket, for consistency :)

claviska commented on src/claviska/SimpleImage.php:
```diff
@@ -842,12 +910,44 @@ public function rotate($angle, $backgroundColor = 'transparent') {
-    imagecolortransparent($this->image, imagecolorallocatealpha($this->image, 0, 0, 0, 127));
+    if ($this->mimeType == 'image/gif' && GD_BUNDLED === 1 && $this->bits == 8){
+        
+      $trnprt_indx = imagecolortransparent($this->image);
```
Use camelCase vars

claviska commented on src/claviska/SimpleImage.php:
```diff
@@ -842,12 +910,44 @@ public function rotate($angle, $backgroundColor = 'transparent') {
+      
+    }
+    
+    else {
```
Same line

claviska commented:
I really appreciate your work on this. I suggested a couple of minor corrections to the code (the project doesn't have a linter config yet).

Since I have to maintain this longterm, I'd really like to see some comments detailing the new conditionals. For example, why do we check for `GD_BUNDLE`? And why do we care how many bits an image uses?

Feel free to be verbose if you need to explain it better (links are OK but could break in the future). This will ensure future PRs don't undo your work and help all maintainers understand what's going on.

Thanks again!

raffaelj commented:
I did the cleanup and added some comments. I can't explain everything. Understanding all these gd functions would require a much bigger test setup with single functions and tons of test images. But I hope, I explained it good enough.

There was another stupid, but small bug in my code:

`imagescale($this->image, $width, $height);` must be `$this->image = imagescale($this->image, $width, $height);`

In the screenshots the jpeg files didn't resize anymore. Now it's fixed.

I moved the part for the rotate function into [a new branch](https://github.com/raffaelj/SimpleImage/tree/rotate-transp-gif-bundled). I'm close... but some images work, others are half-filled with artefacts... I won't repair this function. If I ever want to rotate transparent gif files with PHP in non 90° steps, I'll have a look at ImageMagick or I'll simply use CSS to rotate these damn things in the browser.

raffaelj commented:
Don't merge it. I found some more issues. My current fix loses transparency in black image parts of gifs.

adaniello commented:
@raffaelj no news about your PR? Thanks

raffaelj commented:
@adaniello Sorry, I gave up on this. It took me ages to understand the most things and each approach broke another thing. This was my conclusion:
https://github.com/claviska/SimpleImage/issues/236#issuecomment-443920493

And a quote from claviska:

> GD can definitely be very frustrating, which is why I wrote SimpleImage in the first place. This issue, however, has proven impossible to perfect — thanks for trying and sharing your experience!

However, I am using my modified fork on some production setups and it seems to work as long as you don't try to apply filters on transparent images. This is an example, how I used the modified version in Cockpit CMS:
https://github.com/raffaelj/cockpit_SimpleImageFixBlackBackgrounds

adaniello commented:
@raffaelj Thanks for your feedback and your time

claviska commented:
Closing due to lack of activity.
